odoo/api.py

```diff
@@ -797,6 +797,25 @@ def get_values(self, records, field):
             except KeyError:
                 pass
 
+    def get_records_matching(self, records, field, predicate):
+        """ Return the subset of ``records`` where the value of ``field``
+        satisfies some predicate.
+        """
+        field_cache = self._data[field]
+        if field.depends_context:
+            def match(record_id):
+                return any(
+                    predicate(val)
+                    for val in field_cache.get(record_id, {}).values()
+                )
+        else:
+            def match(record_id):
+                try:
+                    return predicate(field_cache[record_id])
```

Comment:
the key of `field_cache` ids of the parent model. You want to check if `record_id` is in the values of `field_cache`, not the keys